project_key/models/project_task.py

```diff
@@ -46,7 +47,13 @@ def create(self, vals):
 
         if project_id:
             project = self.env['project.project'].browse(project_id)
-            vals['key'] = project.get_next_task_key()
+            key = vals.get('key', False)
+            if key:
+                seq = int(key.split("-")[1])
+                if seq >= project.get_task_key_sequence():
+                    project.set_task_key_sequence(seq + 1)
```

Comment:
Wouldn't this create holes in numbering sequence?

Reply:
It would create gap, but imagine that you have to import tasks that already have a key, since key is what people remember, and refer to, they should not be changed (There is a Project Git module which parses commit message looking for key, and links commit to a task, and changing keys would introduce confusion). By increasing task sequence I am making sure that non of the previously assigned task keys is reused. Those holes are made only when creating tasks with keys, which could happen when importing, so in regular use case this would not happen.

Reply:
So it's for import-only? Can we somehow detect that? Since there may be the case when I'm manually creating A-1 A-2 and then mistype and create A-20, as a result next in sequence would be A-21

I don't feel that's expected behaviour

Reply:
Why would you ever type a key? It should be generated, and in default view, it's readonly. I only see need for specifying a key when importing or with RPC. But think it this way. If you always generate keys in sequence, and you have A-20, there must have been A-1...A-19 at some point of time. I am thinking of keys as unique ID, just like in database.

Reply:
I don't see this as a problem really. 

> @alexey-pelykh I have found problems in v12 with this modules... This Improvements are needed in v12 too. Regards. It lools like @agajic-modoolar is quite busy.

I am, indeed :) I like to spend time with my baby :)

Reply:
Do you think it would be good to make that optional?

Reply:
@alexey-pelykh I agree with your suggestion to make this optional. What do you think about adding some context flag for this?

Reply:
Can you provide a use case where problem could emerge with this implementation?

Reply:
1) Not all keys have XXX-1234 format, I think it would be better to consume all digits from end of string, with regexp maybe?
2) This changes current behaviour, without any warning in readme or option to turn it off (maybe make it optional and described?)